### ISSUE TITLE
The issue is modifyM

### ISSUE THREAD

**Issue by nybble41** [andreasabel/shoot-yourself-in-the-foot-with-lenses#1]
The article says: 

>Our first suspicion goes to modifyM. It gets the state, modifies it through l f and writes back the result… Seems exactly what we want it to do.

However that is *not* what you want to do. You're getting the value, *running an arbitrary monadic action in the MonadState*, and then overwriting the *new* state with an updated version of the *old* state. This is an error whether or not lenses are involved, much as of you had written something like: 

```haskell
data Counted { _count :: Int, ... other fields ... }
countM :: MonadState Counted m => m () -> m ()
countM f = do
  s <- get
  f
  put $ s { _count = _count s + 1 }
```

No lenses in sight, but this has exactly the same problem: the provided action can update the state in ways this code doesn't account for and those changes (to `_count` or any other field) are discarded by the `put` at the end based on the original state. 

I'm not exactly sure how to fix this in a general way. Ideally the `MonadState` would somehow be removed from the effect stack provided to the action passed to `modifyM` (or perhaps replaced with a `MonadReader`) so that the state can't be changed concurrently in two places. That would require a method and associated type in the `MonadState` class to provide an equivalent monad stack without the state, e.g. `statelessly :: (MonadState s m, WithoutState s m n) => n r -> m r`. But MTL doesn't really support this currently, and of course it just prevents the examples in the article from working anyway. You couldn't use your effectful lens with the new `modifyM` since it depends on the `MonadState`.

**Comment by andreasabel** [comment]
Yes indeed, one would need a more fine-grained effect type system when one can say that certain effects are disjoint.  In your example, one would like to say that the effects of `f` do not touch the `_count` field of the state.
